Userland/Libraries/LibWeb/DOM/Element.cpp

```diff
@@ -2238,87 +2286,112 @@ Element::Directionality Element::directionality() const
                 return Directionality::Rtl;
         }
 
-        // Otherwise, if the element's value is not the empty string, or if the element is a document element,
-        // the directionality of the element is 'ltr'.
-        if (!value.is_empty() || is_document_element()) {
+        // 2. If element's value is not the empty string, then return 'ltr'.
+        if (value.is_empty())
             return Directionality::Ltr;
-        }
-        // Otherwise, the directionality of the element is the same as the element's parent element's directionality.
-        else {
-            return parent_element()->directionality();
-        }
+
+        // 3. Return null.
+        return {};
     }
 
-    // -> If the element's dir attribute is in the auto state
-    // FIXME: -> If the element is a bdi element and the dir attribute is not in a defined state
-    //    (i.e. it is not present or has an invalid value)
-    if (dir == Dir::Auto) {
-        // Find the first character in tree order that matches the following criteria:
-        // - The character is from a Text node that is a descendant of the element whose directionality is being determined.
-        // - The character is of bidirectional character type L, AL, or R. [BIDI]
-        // - The character is not in a Text node that has an ancestor element that is a descendant of
-        //   the element whose directionality is being determined and that is either:
-        //   - FIXME: A bdi element.
-        //   - A script element.
-        //   - A style element.
-        //   - A textarea element.
-        //   - An element with a dir attribute in a defined state.
-        Optional<u32> found_character;
-        Optional<Unicode::BidirectionalClass> found_character_bidi_class;
-        for_each_in_subtree_of_type<Text>([&](Text const& text_node) {
-            // Discard not-allowed ancestors
-            for (auto* ancestor = text_node.parent(); ancestor && ancestor != this; ancestor = ancestor->parent()) {
-                if (is<HTML::HTMLScriptElement>(*ancestor) || is<HTML::HTMLStyleElement>(*ancestor) || is<HTML::HTMLTextAreaElement>(*ancestor))
-                    return IterationDecision::Continue;
-                if (ancestor->is_element()) {
-                    auto ancestor_element = static_cast<Element const*>(ancestor);
-                    if (ancestor_element->dir().has_value())
-                        return IterationDecision::Continue;
-                }
-            }
+    // 2. If element is a slot element whose root is a shadow root and element's assigned nodes are not empty:
```

Comment:
Does this need a FIXME to check the shadow root/assigned node parts? This branch always returns so I think this will deviate from the spec by not doing so.

Reply:
Looks like I just didn't implement part of the spec text :sweat_smile: . I've updated it so that the implementation matches the text.